### ISSUE TITLE
slot default children rendered even when slot is filled

### ISSUE THREAD

**Issue by y-nk** [withastro/astro#8038]
### What version of `astro` are you using?

latest

### Are you using an SSR adapter? If so, which one?

no

### What package manager are you using?

npm

### What operating system are you using?

macOS

### What browser are you using?

stackblitz

### Describe the Bug

In an astro component, when writing `<slot><SomeFallback /></slot>`, `<SomeFallback />` is rendered for no reason when the slot is filled by the parent.

### What's the expected result?

`<SomeFallback />` should only be rendered when `<slot />` is undefined/empty/not-passed.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/withastro-astro-xj7ddy?file=src/env.d.ts,src/pages/index.astro,src/components/Parent.astro

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by y-nk** [comment]
### Workaround

```diff
- <slot><SomeFallback /></slot>
+ <slot>{Astro.slots.has('default') && <SomeFallback />}</slot>
```

...but it's ugly in userland.

**Comment by matthewp** [comment]
Hm, your Stackblitz is different from the description and the Stackblitz is doing what I would expect. The page passes to the slot and that is what is rendered and the fallback is not. Can you update either the stackblitz to match the example you're describing in this issue or update the issue to match what's in the stackblitz?

**Comment by y-nk** [comment]
@matthewp i've updated the stackblitz, but i think the issue was present in the previous iteration. The example was:

1. in the Parent:

```html
<slot>
  <Child foo="bar" />
</slot>
```

2. in the child:

```md
---
console.log('child render', Astro.props)
---
```

at use in the index page:

```html
<Parent>hello world 1</Parent>
```

From this ↑ my understanding is that `Parent.slots.has('default')` is `true` therefore `<Child />` will not be rendered on screen, but still the log call from its frontmatter shows in the console.

**Comment by lilnasy** [comment]
It seems to be working how you describe it should: `Child.astro` only gets used when a slot is not provided to `Parent`.
<img width="800" alt="image" src="https://github-production-user-asset-6210df.s3.amazonaws.com/69170106/278528298-97b915f0-1d28-4b77-b162-25450843b8af.png">

**Comment by y-nk** [comment]
@lilnasy i agree, but i think it's not proper that the frontmatter part of `Child.astro` still runs even tho the component is not used/rendered.

**Comment by lilnasy** [comment]
Yes, it is because of the eager call.

**Comment by lilnasy** [comment]
- Fixed by #8187
